src/java.base/share/classes/jdk/internal/foreign/abi/CallingSequence.java

```diff
@@ -68,41 +86,117 @@ public List<Binding> returnBindings() {
         return returnBindings;
     }
 
-    public String asString() {
-        StringBuilder sb = new StringBuilder();
+    public boolean forUpcall() {
+        return forUpcall;
+    }
 
-        sb.append("CallingSequence: {\n");
-        sb.append("  MethodType: ").append(mt);
-        sb.append("  FunctionDescriptor: ").append(desc);
-        sb.append("  Argument Bindings:\n");
-        for (int i = 0; i < mt.parameterCount(); i++) {
-            sb.append("    ").append(i).append(": ").append(argumentBindings.get(i)).append("\n");
-        }
-        if (mt.returnType() != void.class) {
-            sb.append("    ").append("Return: ").append(returnBindings).append("\n");
-        }
-        sb.append("}\n");
+    public boolean forDowncall() {
+        return !forUpcall;
+    }
 
-        return sb.toString();
+    /**
+     * Returns the caller method type, which is the high-level method type
+     * for downcalls (the type of the downcall method handle)
+     * and the low-level method type (all primitives, VM facing) for upcalls.
+     *
+     * Note that for downcalls a single parameter in this method type corresponds
+     * to a single argument binding recipe in this calling sequence, but it may
+     * correspond to multiple parameters in the callee method type (for instance
+     * if a struct is split into multiple register values).
+     *
+     * @return the caller method type.
+     */
+    public MethodType callerMethodType() {
```

Comment:
I don't think renaming them to down/upcallType makes sense, since both down/up calls use both the caller and callee method type. There can also be things that make these not equivalent to downcallType/upcallType, such as the SysV need to pass the number of floating point arguments for variadic calls as well: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/foreign/abi/x64/sysv/CallArranger.java#L109-L111

Reply:
I've left this as is for now.